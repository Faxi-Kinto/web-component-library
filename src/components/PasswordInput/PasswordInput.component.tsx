import React, { useState } from 'react';
import * as Styled from './PasswordInput.styles';
import { InputProps } from '../Input/Input.component';
import Icon from '../Icon';
import Input from '../Input';
import { theme } from '../../config';
import { FieldProps } from '../../helpers/FieldProps';
import useField from '../../hooks/useField';

/**
 * @name PasswordInput
 *
 * Password molecule contains of input atom and icons
 *
 * @returns {JSX}
 */

type ValueOnChange = (event: React.ChangeEvent<HTMLInputElement>) => void;

export type PasswordInputProps = FieldProps<string, ValueOnChange> & {
  error?: string;
} & InputProps;

function handleCapsLock(e: React.KeyboardEvent<HTMLInputElement>): boolean {
  //Take char on keypress
  const s = String.fromCharCode(e.which);

  //Check if it is uppercased
  if (s.toUpperCase() === s && s.toLowerCase() !== s && !e.shiftKey) {
    return true;
  } else {
    return false;
  }
}

const PasswordInput: React.FC<PasswordInputProps> = (
  props: PasswordInputProps
): JSX.Element => {
  const { error, label, dirty, touched, ...rest } = props;
  const [inputType, setInputType] = useState<string>('password');
  const [capslook, setCapslook] = useState<boolean>(false);
  const showError = useField(props);
  const showErrorMsg = showError && error !== 'rule-error';

  return (
    <Styled.Container>
      {label && <label>{label}</label>}
      <Styled.InputWrapper>
        <Input
          type={inputType}
          error={showError}
          {...rest}
          onHandleCapsLock={(e: React.KeyboardEvent<HTMLInputElement>) => {
            setCapslook(handleCapsLock(e));
          }}
        />
        <div
          onClick={() => {
            if (inputType === 'password') {
              setInputType('text');
            } else {
              setInputType('password');
            }
          }}
        >
          {inputType === 'password' ? (
            <Styled.EyeIconWrapper>
              <Icon name="eye" color={theme.pallet.DISABLED} />
            </Styled.EyeIconWrapper>
          ) : (
            <Styled.EyeIconWrapper>
              <Icon name="eye-slash" color={theme.pallet.DISABLED} />
            </Styled.EyeIconWrapper>
          )}
        </div>
        {capslook && (
          <Styled.CapslookIconWrapper>
            <Icon name="caps-lock" color={theme.pallet.GRAY} />
          </Styled.CapslookIconWrapper>
        )}
      </Styled.InputWrapper>
      {showErrorMsg}
    </Styled.Container>
  );
};

export default PasswordInput as React.FC<PasswordInputProps>;
