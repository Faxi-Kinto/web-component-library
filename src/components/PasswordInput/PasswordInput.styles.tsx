import styled from 'styled-components';
import { positionAbsolute, pxToRem, fontSize } from '../../styles/basics';
import { theme } from '../../config';

export const Container = styled.div`
  width: 100%;

  > label {
    color: ${theme.pallet.GRAY};
    display: flex;
    height: ${pxToRem('24px')};
    align-items: flex-end;
    margin-bottom: ${pxToRem('5px')};
    ${fontSize('16px')}
  }
`;

export const InputWrapper = styled.div`
  position: relative;
  input {
    padding-right: ${pxToRem('70px')};
  }
`;

export const EyeIconWrapper = styled.div`
  ${positionAbsolute(pxToRem('14px'), pxToRem('28px'), '', '')};
`;

export const CapslookIconWrapper = styled.div`
  ${positionAbsolute(pxToRem('14px'), pxToRem('16px'), '', '')};
`;
