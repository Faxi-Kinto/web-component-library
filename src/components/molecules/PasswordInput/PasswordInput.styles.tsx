import styled from 'styled-components';
import { positionAbsolute } from 'styles/basics';
import { theme } from 'config';
import { fontSize } from 'styles/fonts';

export const Container = styled.div`
  width: 100%;

  > label {
    color: ${theme.pallet.GRAY};
    display: flex;
    height: 24px;
    align-items: flex-end;
    margin-bottom: 5px;
    ${fontSize('16px')}
  }
`;

export const InputWrapper = styled.div`
  position: relative;
  input {
    padding-right: 70px;
  }
`;

export const EyeIconWrapper = styled.div`
  ${positionAbsolute('14px', '37px', '', '')};
`;

export const CapslookIconWrapper = styled.div`
  ${positionAbsolute('14px', '16px', '', '')};
`;
