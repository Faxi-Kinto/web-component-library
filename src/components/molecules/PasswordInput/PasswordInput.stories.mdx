import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";

import PasswordInput from "./";

<Meta title="Core|Molecules/PasswordInput" />

# PasswordInput Component

Straithforward Password component.

It contains an eye icon. When you click on the eye icon, the text will be shown.

There is also a caps-lock icon. When caps-lock on your keyboard is pressed, icon will be shown.

When an error occured its border become red.

## Implementation

#### Default
<Preview>
  <Story name="default">
    <PasswordInput />
  </Story>
</Preview>

#### With Error
<Preview>
  <Story name="With an error">
    <PasswordInput error="Error message" touched={true} dirty={true} />
  </Story>
</Preview>

## PasswordInput Props

<Props of={PasswordInput} />

## Notes

- Error prop is required. Default error value is false.
