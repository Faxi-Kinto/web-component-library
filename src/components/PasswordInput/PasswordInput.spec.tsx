import React from 'react';
import renderer from 'react-test-renderer';
import PasswordInput from './PasswordInput.component';

test('PasswordInput renders correctly', () => {
  const component = renderer.create(<PasswordInput />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
