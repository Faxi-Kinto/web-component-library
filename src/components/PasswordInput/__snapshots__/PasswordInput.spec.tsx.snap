// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput renders correctly 1`] = `
.c0 {
  width: 100%;
}

.c0 > label {
  color: #4E606A;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 1.5rem;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-bottom: 0.3125rem;
  font-size: 1rem;
  line-height: 16px;
}

.c1 {
  position: relative;
}

.c1 input {
  padding-right: 4.375rem;
}

.c3 {
  position: absolute;
  top: 0.875rem;
  right: 1.75rem;
}

.c2 {
  width: 100%;
}

.c2 > * {
  margin: 0 0 0.625rem 0;
}

.c2 > :last-child {
  margin: 0;
}

.c2 > label {
  color: #4E606A;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 24px;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-bottom: 5px;
  font-size: 1rem;
  line-height: 16px;
}

.c2 > input {
  width: 100%;
  height: 2.8125rem;
  padding: 0.5rem 1rem 0.5rem 1rem;
  box-shadow: 0.03125rem 0.03125rem 0.078125rem #00000029;
  border: 0.0625rem solid #95A1A6;
  outline: none;
}

.c2 > input.input--error {
  border: 0.0625rem solid #EB4820;
}

.c2 > input:focus,
.c2 > input.input--error {
  box-shadow: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1rem;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <input
        className=""
        onChange={[Function]}
        onKeyPress={[Function]}
        type="password"
        value=""
      />
    </div>
    <div
      onClick={[Function]}
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-eye fa-w-18 fa-1x "
            color="#95A1A6"
            data-icon="eye"
            data-prefix="far"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M288 144a110.94 110.94 0 0 0-31.24 5 55.4 55.4 0 0 1 7.24 27 56 56 0 0 1-56 56 55.4 55.4 0 0 1-27-7.24A111.71 111.71 0 1 0 288 144zm284.52 97.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400c-98.65 0-189.09-55-237.93-144C98.91 167 189.34 112 288 112s189.09 55 237.93 144C477.1 345 386.66 400 288 400z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
